### ISSUE TITLE
[JENKINS-38268] Parallel step and closure scope

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/workflow-cps-plugin#1221]
<p>I'm experiencing some odd behaviour with the parallel step related to variable scoping. The following minimal pipeline script demonstrates my problem.</p>

<pre>
def fn = { val -> println val }

parallel([
  a: { fn(<span class="code-quote">'a'</span>) },
  b: { fn(<span class="code-quote">'b'</span>) }
])
</pre>
<p>Expected output</p>

<pre>
a
b
</pre>
<p>(or b then a, order of execution should be undefined)</p>

<p>Actual output</p>

<pre>
b
b
</pre>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=philgrayson">philgrayson</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268" target="_blank">Parallel step and closure scope</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer">abayer</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>component(s)</b>: workflow-cps-plugin
<li><b>label(s)</b>: parallel
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-06-13T20:48:43+00:00
<li><b>votes</b>: 16
<li><b>watchers</b>: 27
<li><b>imported</b>: 2025-12-07
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>I'm experiencing some odd behaviour with the parallel step related to variable scoping. The following minimal pipeline script demonstrates my problem.</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def fn = { val -&gt; println val }

parallel([
  a: { fn(<span class="code-quote">'a'</span>) },
  b: { fn(<span class="code-quote">'b'</span>) }
])
</pre>
</div></div>
<p>Expected output</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">a
b
</pre>
</div></div>
<p>(or b then a, order of execution should be undefined)</p>

<p>Actual output</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">b
b
</pre>
</div></div></pre>
</details>
</details>
<details><summary><i>environment</i></summary>

```
Jenkins: 2.7.4 (affecting 1.6 too)<br/>
Pipeline: 2.3 (tested on fresh install with latest dependencies)<br/>
Environment setup;<br/>
1. Run &quot;docker run -p 8080:8080 -p 50000:50000 jenkins&quot;<br/>
2. Browse to <a href="http://localhost:8080">http://localhost:8080</a><br/>
3. Perform initial jenkins setup - install &#39;pipeline&#39; plugin
```
</details>
<details><summary><i>4 attachments</i></summary>

- [console_output.png](https://issues.jenkins.io/secure/attachment/33973/console_output.png)
  > ![console_output.png](https://issues.jenkins.io/secure/attachment/33973/console_output.png)
- [console_output_no_node_step.png](https://issues.jenkins.io/secure/attachment/33971/console_output_no_node_step.png)
  > ![console_output_no_node_step.png](https://issues.jenkins.io/secure/attachment/33971/console_output_no_node_step.png)
- [pipeline_step.png](https://issues.jenkins.io/secure/attachment/33972/pipeline_step.png)
  > ![pipeline_step.png](https://issues.jenkins.io/secure/attachment/33972/pipeline_step.png)
- [pipeline_steps.png](https://issues.jenkins.io/secure/attachment/36077/pipeline_steps.png)
  > ![pipeline_steps.png](https://issues.jenkins.io/secure/attachment/36077/pipeline_steps.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=273278&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-273278">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I doubt this is a duplicate of <a href="https://issue-redirect.jenkins.io/issue/26481" title="Mishandling of binary methods accepting Closure" class="issue-link" data-issue-key="JENKINS-26481"><del>JENKINS-26481</del></a>. <tt>call</tt> in this case should refer to the implicit method defined in CPS-transformed sources, not a binary method. And it is the <em>receiver</em>, not an <em>argument</em>, which is a closure.</p>
</pre>
</details>

</ul>
</details>

<p>I doubt this is a duplicate of <a href="https://issue-redirect.jenkins.io/issue/26481" title="Mishandling of binary methods accepting Closure" class="issue-link" data-issue-key="JENKINS-26481"><del>JENKINS-26481</del></a>. <tt>call</tt> in this case should refer to the implicit method defined in CPS-transformed sources, not a binary method. And it is the <em>receiver</em>, not an <em>argument</em>, which is a closure.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>aleol57:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=274767&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-274767">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Been able to replicate this consistently without having to add a node block, in fact I still haven't found a way to not replicate it.</p>

<p>Composited closures also seem to be even more unhappy about being parallelized.</p>

<p>Script:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def heavyweight_task = { fn-&gt;
    println <span class="code-quote">"Doing heavy work on ${fn}..."</span>
}

def work = [
    file10: heavyweight_task.curry(<span class="code-quote">'file10'</span>),
    file11: heavyweight_task.curry(<span class="code-quote">'file11'</span>),
    file20: ({<span class="code-quote">'file20'</span>} &gt;&gt; heavyweight_task),
    file21: ({<span class="code-quote">'file21'</span>} &gt;&gt; heavyweight_task),
    file30: { heavyweight_task.call(<span class="code-quote">'file30'</span>) },
    file31: { heavyweight_task.call(<span class="code-quote">'file31'</span>) },
    file40: heavyweight_task.clone().curry(<span class="code-quote">'file40'</span>),
    file41: heavyweight_task.clone().curry(<span class="code-quote">'file41'</span>),
    file50: heavyweight_task.dehydrate().curry(<span class="code-quote">'file50'</span>),
    file51: heavyweight_task.dehydrate().curry(<span class="code-quote">'file51'</span>),
]

parallel(work)
</pre>
</div></div>

<p>output:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[Pipeline] parallel
[Pipeline] [file10] { (Branch: file10)
[Pipeline] [file11] { (Branch: file11)
[Pipeline] [file20] { (Branch: file20)
[Pipeline] [file21] { (Branch: file21)
[Pipeline] [file30] { (Branch: file30)
[Pipeline] [file31] { (Branch: file31)
[Pipeline] [file40] { (Branch: file40)
[Pipeline] [file41] { (Branch: file41)
[Pipeline] [file50] { (Branch: file50)
[Pipeline] [file51] { (Branch: file51)
[Pipeline] [file10] echo
[file10] Doing heavy work on file51...
[Pipeline] [file10] }
[Pipeline] [file11] echo
[file11] Doing heavy work on file51...
[Pipeline] [file11] }
[Pipeline] [file20] }
[Pipeline] [file21] }
[Pipeline] [file40] echo
[file40] Doing heavy work on file31...
[Pipeline] [file40] }
[Pipeline] [file41] echo
[file41] Doing heavy work on file31...
[Pipeline] [file41] }
[Pipeline] [file50] echo
[file50] Doing heavy work on file31...
[Pipeline] [file50] }
[Pipeline] [file51] echo
[file51] Doing heavy work on file31...
[Pipeline] [file51] }
[Pipeline] [file30] echo
[file30] Doing heavy work on file31...
[Pipeline] [file30] }
[Pipeline] [file31] echo
[file31] Doing heavy work on file31...
[Pipeline] [file31] }
[Pipeline] <span class="code-comment">// parallel
</span>[Pipeline] End of Pipeline
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Been able to replicate this consistently without having to add a node block, in fact I still haven't found a way to not replicate it.</p>

<p>Composited closures also seem to be even more unhappy about being parallelized.</p>

<p>Script:</p>

<pre>
def heavyweight_task = { fn->
    println <span class="code-quote">"Doing heavy work on ${fn}..."</span>
}

def work = [
    file10: heavyweight_task.curry(<span class="code-quote">'file10'</span>),
    file11: heavyweight_task.curry(<span class="code-quote">'file11'</span>),
    file20: ({<span class="code-quote">'file20'</span>} >> heavyweight_task),
    file21: ({<span class="code-quote">'file21'</span>} >> heavyweight_task),
    file30: { heavyweight_task.call(<span class="code-quote">'file30'</span>) },
    file31: { heavyweight_task.call(<span class="code-quote">'file31'</span>) },
    file40: heavyweight_task.clone().curry(<span class="code-quote">'file40'</span>),
    file41: heavyweight_task.clone().curry(<span class="code-quote">'file41'</span>),
    file50: heavyweight_task.dehydrate().curry(<span class="code-quote">'file50'</span>),
    file51: heavyweight_task.dehydrate().curry(<span class="code-quote">'file51'</span>),
]

parallel(work)
</pre>

<p>output:</p>

<pre>
[Pipeline] parallel
[Pipeline] [file10] { (Branch: file10)
[Pipeline] [file11] { (Branch: file11)
[Pipeline] [file20] { (Branch: file20)
[Pipeline] [file21] { (Branch: file21)
[Pipeline] [file30] { (Branch: file30)
[Pipeline] [file31] { (Branch: file31)
[Pipeline] [file40] { (Branch: file40)
[Pipeline] [file41] { (Branch: file41)
[Pipeline] [file50] { (Branch: file50)
[Pipeline] [file51] { (Branch: file51)
[Pipeline] [file10] echo
[file10] Doing heavy work on file51...
[Pipeline] [file10] }
[Pipeline] [file11] echo
[file11] Doing heavy work on file51...
[Pipeline] [file11] }
[Pipeline] [file20] }
[Pipeline] [file21] }
[Pipeline] [file40] echo
[file40] Doing heavy work on file31...
[Pipeline] [file40] }
[Pipeline] [file41] echo
[file41] Doing heavy work on file31...
[Pipeline] [file41] }
[Pipeline] [file50] echo
[file50] Doing heavy work on file31...
[Pipeline] [file50] }
[Pipeline] [file51] echo
[file51] Doing heavy work on file31...
[Pipeline] [file51] }
[Pipeline] [file30] echo
[file30] Doing heavy work on file31...
[Pipeline] [file30] }
[Pipeline] [file31] echo
[file31] Doing heavy work on file31...
[Pipeline] [file31] }
[Pipeline] <span class="code-comment">// parallel
</span>[Pipeline] End of Pipeline
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=274789&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-274789">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>No idea offhand, will have to reproduce in a functional test and study in a debugger.</p>
</pre>
</details>

</ul>
</details>

<p>No idea offhand, will have to reproduce in a functional test and study in a debugger.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>akovi:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=279350&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-279350">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Seems like the names of the parameters of the called closures play some role. For example:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">
def finallyHandler(param, closure) {
    echo <span class="code-quote">"HANDLER:param=$param"</span>
    closure(param, param)
}

work = [
  <span class="code-quote">"1"</span>: { finallyHandler(1) { p, p1 -&gt; echo <span class="code-quote">"p=$p p1=$p1"</span> } },
  <span class="code-quote">"2"</span>: { finallyHandler(2) { p, p2 -&gt; echo <span class="code-quote">"p=$p p2=$p2"</span> } },
  <span class="code-quote">"3"</span>: { finallyHandler(3) { p, p3 -&gt; echo <span class="code-quote">"p=$p p3=$p3"</span> } },
]

parallel work
</pre>
</div></div>

<p>Output: (we want to see the same value for p= and pX=)</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">[Pipeline] parallel
[Pipeline] [1] { (Branch: 1)
[Pipeline] [2] { (Branch: 2)
[Pipeline] [3] { (Branch: 3)
[Pipeline] [1] echo
<span class="code-quote">"[1] HANDLER:param=1"</span>
[Pipeline] [2] echo
<span class="code-quote">"[2] HANDLER:param=2"</span>
[Pipeline] [3] echo
<span class="code-quote">"[3] HANDLER:param=3"</span>
[Pipeline] [1] echo
<span class="code-quote">"[1] p=3 p1=1"</span>
[Pipeline] [1] }
[Pipeline] [2] echo
<span class="code-quote">"[2] p=3 p2=2"</span>
[Pipeline] [2] }
[Pipeline] [3] echo
<span class="code-quote">"[3] p=3 p3=3"</span>
[Pipeline] [3] }
[Pipeline] <span class="code-comment">// parallel
</span>[Pipeline] End of Pipeline
Finished: SUCCESS
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>Seems like the names of the parameters of the called closures play some role. For example:</p>


<pre>

def finallyHandler(param, closure) {
    echo <span class="code-quote">"HANDLER:param=$param"</span>
    closure(param, param)
}

work = [
  <span class="code-quote">"1"</span>: { finallyHandler(1) { p, p1 -> echo <span class="code-quote">"p=$p p1=$p1"</span> } },
  <span class="code-quote">"2"</span>: { finallyHandler(2) { p, p2 -> echo <span class="code-quote">"p=$p p2=$p2"</span> } },
  <span class="code-quote">"3"</span>: { finallyHandler(3) { p, p3 -> echo <span class="code-quote">"p=$p p3=$p3"</span> } },
]

parallel work
</pre>

<p>Output: (we want to see the same value for p= and pX=)</p>


<pre>
[Pipeline] parallel
[Pipeline] [1] { (Branch: 1)
[Pipeline] [2] { (Branch: 2)
[Pipeline] [3] { (Branch: 3)
[Pipeline] [1] echo
<span class="code-quote">"[1] HANDLER:param=1"</span>
[Pipeline] [2] echo
<span class="code-quote">"[2] HANDLER:param=2"</span>
[Pipeline] [3] echo
<span class="code-quote">"[3] HANDLER:param=3"</span>
[Pipeline] [1] echo
<span class="code-quote">"[1] p=3 p1=1"</span>
[Pipeline] [1] }
[Pipeline] [2] echo
<span class="code-quote">"[2] p=3 p2=2"</span>
[Pipeline] [2] }
[Pipeline] [3] echo
<span class="code-quote">"[3] p=3 p3=3"</span>
[Pipeline] [3] }
[Pipeline] <span class="code-comment">// parallel
</span>[Pipeline] End of Pipeline
Finished: SUCCESS
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=279443&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-279443">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I wonder if <tt>ClosureCallEnv</tt> is misbehaving. Need to go through this in a debugger.</p>
</pre>
</details>

</ul>
</details>

<p>I wonder if <tt>ClosureCallEnv</tt> is misbehaving. Need to go through this in a debugger.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/externl.jpg" title="externl's avatar" /> externl:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=281322&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281322">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>We experience this too, it's quite annoying and hard to work around.</p>
</pre>
</details>

</ul>
</details>

<p>We experience this too, it's quite annoying and hard to work around.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jlpinardon.png" title="jlpinardon's avatar" /> jlpinardon:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=281985&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-281985">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Same problem when trying to loop over a map of git repositories :
def scmRepo = [
    'jobRepo': [
        'url': "....."
        'branch': "origin/master",
        'credentialsId': '.....',
        'targetDir': 'src'
    ],
    'envRepo': [
        'url': "....."
        'branch': "origin/master",
        'credentialsId': '.....',
        'targetDir': 'src'
    ],
    'srcRepo': [
        'url': "....."
        'branch': "origin/master",
        'credentialsId': '.....',
        'targetDir': 'src'
    ]
]
@NonCPS
def runCheckOut(scmRepo) {
    def branches = <span class="error">&#91;: &#93;</span>
    scmRepo.each() {
        repoName,
        repoRef - &gt;
        branches<span class="error">&#91;&quot;${repoName}&quot;&#93;</span> = {
            node {
                println "\nCheckout content from ${repoRef.url}"
                //your logic there
            }
        }
    }
    parallel branches
}</p>

<p>When using runCheckOut, it runs but never ends. No error message is displayed.</p>

</pre>
</details>

</ul>
</details>

<p>Same problem when trying to loop over a map of git repositories :<br/>
def scmRepo = [<br/>
    'jobRepo': [<br/>
'url': "....."<br/>
'branch': "origin/master",<br/>
'credentialsId': '.....',<br/>
'targetDir': 'src'<br/>
    ],<br/>
    'envRepo': [<br/>
'url': "....."<br/>
'branch': "origin/master",<br/>
'credentialsId': '.....',<br/>
'targetDir': 'src'<br/>
    ],<br/>
    'srcRepo': [<br/>
'url': "....."<br/>
'branch': "origin/master",<br/>
'credentialsId': '.....',<br/>
'targetDir': 'src'<br/>
    ]<br/>
]<br/>
@​NonCPS<br/>
def runCheckOut(scmRepo) {<br/>
    def branches = <span class="error">&#91;: &#93;</span><br/>
    scmRepo.each() {<br/>
repoName,<br/>
repoRef - ><br/>
branches<span class="error">&#91;"${repoName}"&#93;</span> = {<br/>
    node {<br/>
println "\nCheckout content from ${repoRef.url}"<br/>
//your logic there<br/>
    }<br/>
}<br/>
    }<br/>
    parallel branches<br/>
}</p>

<p>When using runCheckOut, it runs but never ends. No error message is displayed.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nvgoldin:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=288834&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288834">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi, I'm experiencing what I think is a symptom of the same issue, while trying to trigger parallel builds, take this pipeline job, assuming the dummy jobs 'job-1,2,3' exists:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">@NonCPS
def get_dummy_params(val)
{
    <span class="code-keyword">return</span> [string(name: <span class="code-quote">'dummy'</span>, value: <span class="code-quote">"$val"</span>)]
}

@NonCPS
def create_jobs()
{
    def jobs = [:]
    (1..3).each { jobs[<span class="code-quote">"job-$it"</span>] = { -&gt; build([job: <span class="code-quote">"job-$it"</span>, parameters: get_dummy_params(it) ]) } }
    <span class="code-keyword">return</span> jobs
}

stage (<span class="code-quote">'triggering'</span>) {
    parallel(create_jobs())
}
</pre>
</div></div>

<p>I would expect it to trigger the 3 jobs in parallel, instead it  'nests' under 'job-1' branch all 3 jobs, and creates empty branches for 'job-2' and 'job-3', causing the job to never end. Console output:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Started by user admin
[Pipeline] stage
[Pipeline] { (triggering)
[Pipeline] parallel
[Pipeline] [job-1] { (Branch: job-1)
[Pipeline] [job-1] build (Building job-1)
[job-1] Scheduling project: job-1
[Pipeline] [job-2] { (Branch: job-2)
[Pipeline] [job-1] build (Building job-2)
[job-1] Scheduling project: job-2
[Pipeline] [job-3] { (Branch: job-3)
[Pipeline] [job-1] build (Building job-3)
[job-1] Scheduling project: job-3 !pipeline_steps.png|thumbnail! 
[job-1] Starting building: job-3 #1
...
</pre>
</div></div>

<p>I found only 2 workarounds:
1. Don't use any functions to generate data fed into 'parallel' function(i.e. create everything under the 'stage' step). 
2. Feeding a newly created map to parallel and redefining the groovy closures with a '.call()' (frankly I have no idea how I came up with this). 
I.e. replace the 'stage' in the previous example with:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">stage (<span class="code-quote">'triggering'</span>) {
    def jobs = create_jobs()
    parallel([
        <span class="code-quote">'job-1'</span>: { jobs[<span class="code-quote">'job-1'</span>].call() },
        <span class="code-quote">'job-2'</span>: { jobs[<span class="code-quote">'job-2'</span>].call() },
        <span class="code-quote">'job-3'</span>: { jobs[<span class="code-quote">'job-3'</span>].call() },
        ])
}
</pre>
</div></div>

<p>Output as expected:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">Started by user admin
[Pipeline] stage
[Pipeline] { (triggering)
[Pipeline] parallel
[Pipeline] [job-1] { (Branch: job-1)
[Pipeline] [job-2] { (Branch: job-2)
[Pipeline] [job-3] { (Branch: job-3)
[Pipeline] [job-1] build (Building job-1)
[job-1] Scheduling project: job-1
[Pipeline] [job-2] build (Building job-2)
[job-2] Scheduling project: job-2
[Pipeline] [job-3] build (Building job-3)
[job-3] Scheduling project: job-3
[job-1] Starting building: job-1 #2
[job-2] Starting building: job-2 #2
[Pipeline] [job-2] }
[job-3] Starting building: job-3 #2
[Pipeline] [job-1] }
[Pipeline] [job-3] }
[Pipeline] <span class="code-comment">// parallel
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// stage
</span>[Pipeline] End of Pipeline
Finished: SUCCESS
</pre>
</div></div>

<p>Though both workarounds don't allow code to be reused, as the list of jobs has to be generated inside the 'stage'.
Tested on Jenkins ver. 2.32.2, using the docker image.</p>
</pre>
</details>

</ul>
</details>

<p>Hi, I'm experiencing what I think is a symptom of the same issue, while trying to trigger parallel builds, take this pipeline job, assuming the dummy jobs 'job-1,2,3' exists:</p>

<pre>
@​NonCPS
def get_dummy_params(val)
{
    <span class="code-keyword">return</span> [string(name: <span class="code-quote">'dummy'</span>, value: <span class="code-quote">"$val"</span>)]
}

@​NonCPS
def create_jobs()
{
    def jobs = [:]
    (1..3).each { jobs[<span class="code-quote">"job-$it"</span>] = { -> build([job: <span class="code-quote">"job-$it"</span>, parameters: get_dummy_params(it) ]) } }
    <span class="code-keyword">return</span> jobs
}

stage (<span class="code-quote">'triggering'</span>) {
    parallel(create_jobs())
}
</pre>

<p>I would expect it to trigger the 3 jobs in parallel, instead it  'nests' under 'job-1' branch all 3 jobs, and creates empty branches for 'job-2' and 'job-3', causing the job to never end. Console output:</p>

<pre>
Started by user admin
[Pipeline] stage
[Pipeline] { (triggering)
[Pipeline] parallel
[Pipeline] [job-1] { (Branch: job-1)
[Pipeline] [job-1] build (Building job-1)
[job-1] Scheduling project: job-1
[Pipeline] [job-2] { (Branch: job-2)
[Pipeline] [job-1] build (Building job-2)
[job-1] Scheduling project: job-2
[Pipeline] [job-3] { (Branch: job-3)
[Pipeline] [job-1] build (Building job-3)
[job-1] Scheduling project: job-3 !pipeline_steps.png|thumbnail! 
[job-1] Starting building: job-3 #1
...
</pre>

<p>I found only 2 workarounds:<br/>
1. Don't use any functions to generate data fed into 'parallel' function(i.e. create everything under the 'stage' step). <br/>
2. Feeding a newly created map to parallel and redefining the groovy closures with a '.call()' (frankly I have no idea how I came up with this). <br/>
I.e. replace the 'stage' in the previous example with:</p>

<pre>
stage (<span class="code-quote">'triggering'</span>) {
    def jobs = create_jobs()
    parallel([
<span class="code-quote">'job-1'</span>: { jobs[<span class="code-quote">'job-1'</span>].call() },
<span class="code-quote">'job-2'</span>: { jobs[<span class="code-quote">'job-2'</span>].call() },
<span class="code-quote">'job-3'</span>: { jobs[<span class="code-quote">'job-3'</span>].call() },
])
}
</pre>

<p>Output as expected:</p>

<pre>
Started by user admin
[Pipeline] stage
[Pipeline] { (triggering)
[Pipeline] parallel
[Pipeline] [job-1] { (Branch: job-1)
[Pipeline] [job-2] { (Branch: job-2)
[Pipeline] [job-3] { (Branch: job-3)
[Pipeline] [job-1] build (Building job-1)
[job-1] Scheduling project: job-1
[Pipeline] [job-2] build (Building job-2)
[job-2] Scheduling project: job-2
[Pipeline] [job-3] build (Building job-3)
[job-3] Scheduling project: job-3
[job-1] Starting building: job-1 #2
[job-2] Starting building: job-2 #2
[Pipeline] [job-2] }
[job-3] Starting building: job-3 #2
[Pipeline] [job-1] }
[Pipeline] [job-3] }
[Pipeline] <span class="code-comment">// parallel
</span>[Pipeline] }
[Pipeline] <span class="code-comment">// stage
</span>[Pipeline] End of Pipeline
Finished: SUCCESS
</pre>

<p>Though both workarounds don't allow code to be reused, as the list of jobs has to be generated inside the 'stage'.<br/>
Tested on Jenkins ver. 2.32.2, using the docker image.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=288885&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288885">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nvgoldin" class="user-hover" rel="nvgoldin">nvgoldin</a> <tt>create_jobs</tt> is wrong. You may not call steps like <tt>build</tt> from inside a method marked <tt>@NonCPS</tt>.</p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nvgoldin" class="user-hover" rel="nvgoldin">nvgoldin</a> <tt>create_jobs</tt> is wrong. You may not call steps like <tt>build</tt> from inside a method marked <tt>@​NonCPS</tt>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>nvgoldin:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=288904&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-288904">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a>, thanks! I changed the code to simple for loops without iterators , and without '@NonCPS' and it seems to work . A small gotcha I encountered is that the looping variable needs to be aliased, otherwise only the last variable will be sent to all closures(unlike when using iterators, I think). For reference, here is the code that <b>does</b> work:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def get_dummy_params(val)
{
    <span class="code-keyword">return</span> [string(name: <span class="code-quote">'dummy'</span>, value: <span class="code-quote">"$val"</span>)]
}

def create_jobs()
{
    def jobs = [:]
    <span class="code-keyword">for</span> (<span class="code-object">int</span> i=1; i &lt;= 3; i++) {
        def x = i
        jobs[<span class="code-quote">"job-$x"</span>] = { -&gt; build([job: <span class="code-quote">"job-$x"</span>, parameters: get_dummy_params(x) ]) }
    }
    <span class="code-keyword">return</span> jobs
}

stage (<span class="code-quote">'triggering'</span>) {
    parallel(create_jobs())
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a>, thanks! I changed the code to simple for loops without iterators , and without '@​NonCPS' and it seems to work . A small gotcha I encountered is that the looping variable needs to be aliased, otherwise only the last variable will be sent to all closures(unlike when using iterators, I think). For reference, here is the code that <b>does</b> work:</p>

<pre>
def get_dummy_params(val)
{
    <span class="code-keyword">return</span> [string(name: <span class="code-quote">'dummy'</span>, value: <span class="code-quote">"$val"</span>)]
}

def create_jobs()
{
    def jobs = [:]
    <span class="code-keyword">for</span> (<span class="code-object">int</span> i=1; i <= 3; i++) {
def x = i
jobs[<span class="code-quote">"job-$x"</span>] = { -> build([job: <span class="code-quote">"job-$x"</span>, parameters: get_dummy_params(x) ]) }
    }
    <span class="code-keyword">return</span> jobs
}

stage (<span class="code-quote">'triggering'</span>) {
    parallel(create_jobs())
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=289083&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-289083">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote><p>the looping variable needs to be aliased, otherwise only the last variable will be sent to all closures</p></blockquote>

<p>Yes this is just a general aspect of Groovy.</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-none">$ groovy -e 'def closures = []; for (int i = 0; i &lt; 10; i++) {closures += { -&gt; println i}}; closures.each {it()}'
10
10
10
10
10
10
10
10
10
10
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<blockquote><p>the looping variable needs to be aliased, otherwise only the last variable will be sent to all closures</p></blockquote>

<p>Yes this is just a general aspect of Groovy.</p>


<pre>
$ groovy -e 'def closures = []; for (int i = 0; i < 10; i++) {closures += { -> println i}}; closures.each {it()}'
10
10
10
10
10
10
10
10
10
10
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/tknerr.jpg" title="tknerr's avatar" /> tknerr:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=289429&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-289429">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I believe <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nvgoldin" class="user-hover" rel="nvgoldin">nvgoldin</a>'s example is not representative for this issue, since `get_dummy_params` is a method. If it were closure instead (which is totally not needed in this case, so it would be quite artificial) it would fail again, I suspect.</p>

<p>Here is another example which suffers from the issue, and I have totally no clue how to work around it:</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-groovy"><span class="code-keyword">def</span> onEachSlave(doStuff) {
  <span class="code-keyword">def</span> doStuffClosures = [:]
  <span class="code-keyword">for</span> (slave <span class="code-keyword">in</span> [<span class="code-quote">'slavelnx1'</span>, <span class="code-quote">'slavelnx2'</span>, <span class="code-quote">'slavelnx3'</span>]) {
    <span class="code-keyword">def</span> s = slave
    doStuffClosures[s] = { echo <span class="code-quote">"running on ${s}..."</span>; doStuff(s); echo <span class="code-quote">"...done on ${s}!"</span> }
  }
  <span class="code-keyword">return</span> doStuffClosures
}

parallel(onEachSlave { slave -&gt;
    echo <span class="code-quote">"doing stuff on ${slave}..."</span>
})
</pre>
</div></div>


<p>The output is:</p>

<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>[Pipeline] parallel
[Pipeline] [slavelnx1] { (Branch: slavelnx1)
[Pipeline] [slavelnx2] { (Branch: slavelnx2)
[Pipeline] [slavelnx3] { (Branch: slavelnx3)
[Pipeline] [slavelnx1] echo
[slavelnx1] running on slavelnx1...
[Pipeline] [slavelnx2] echo
[slavelnx2] running on slavelnx2...
[Pipeline] [slavelnx3] echo
[slavelnx3] running on slavelnx3...
[Pipeline] [slavelnx1] echo
[slavelnx1] doing stuff on slavelnx3
[Pipeline] [slavelnx1] echo
[slavelnx1] ...done on slavelnx1!
[Pipeline] [slavelnx1] }
[Pipeline] [slavelnx2] echo
[slavelnx2] doing stuff on slavelnx3
[Pipeline] [slavelnx2] echo
[slavelnx2] ...done on slavelnx2!
[Pipeline] [slavelnx2] }
[Pipeline] [slavelnx3] echo
[slavelnx3] doing stuff on slavelnx3
[Pipeline] [slavelnx3] echo
[slavelnx3] ...done on slavelnx3!
[Pipeline] [slavelnx3] }
[Pipeline] // parallel
[Pipeline] End of Pipeline
</pre>
</div></div>

<p>This is Jenkins 2.32.1 and workflow-cps 2.23</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> can you confirm this is the issue, or is it just my missing groovy foo?</p>

<p>Would also be happy to hear if someone can come up with a workaround for this</p>
</pre>
</details>

</ul>
</details>

<p>I believe <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nvgoldin" class="user-hover" rel="nvgoldin">nvgoldin</a>'s example is not representative for this issue, since `get_dummy_params` is a method. If it were closure instead (which is totally not needed in this case, so it would be quite artificial) it would fail again, I suspect.</p>

<p>Here is another example which suffers from the issue, and I have totally no clue how to work around it:</p>


<pre>
<span class="code-keyword">def</span> onEachSlave(doStuff) {
  <span class="code-keyword">def</span> doStuffClosures = [:]
  <span class="code-keyword">for</span> (slave <span class="code-keyword">in</span> [<span class="code-quote">'slavelnx1'</span>, <span class="code-quote">'slavelnx2'</span>, <span class="code-quote">'slavelnx3'</span>]) {
    <span class="code-keyword">def</span> s = slave
    doStuffClosures[s] = { echo <span class="code-quote">"running on ${s}..."</span>; doStuff(s); echo <span class="code-quote">"...done on ${s}!"</span> }
  }
  <span class="code-keyword">return</span> doStuffClosures
}

parallel(onEachSlave { slave ->
    echo <span class="code-quote">"doing stuff on ${slave}..."</span>
})
</pre>


<p>The output is:</p>



```
[Pipeline] parallel
[Pipeline] [slavelnx1] { (Branch: slavelnx1)
[Pipeline] [slavelnx2] { (Branch: slavelnx2)
[Pipeline] [slavelnx3] { (Branch: slavelnx3)
[Pipeline] [slavelnx1] echo
[slavelnx1] running on slavelnx1...
[Pipeline] [slavelnx2] echo
[slavelnx2] running on slavelnx2...
[Pipeline] [slavelnx3] echo
[slavelnx3] running on slavelnx3...
[Pipeline] [slavelnx1] echo
[slavelnx1] doing stuff on slavelnx3
[Pipeline] [slavelnx1] echo
[slavelnx1] ...done on slavelnx1!
[Pipeline] [slavelnx1] }
[Pipeline] [slavelnx2] echo
[slavelnx2] doing stuff on slavelnx3
[Pipeline] [slavelnx2] echo
[slavelnx2] ...done on slavelnx2!
[Pipeline] [slavelnx2] }
[Pipeline] [slavelnx3] echo
[slavelnx3] doing stuff on slavelnx3
[Pipeline] [slavelnx3] echo
[slavelnx3] ...done on slavelnx3!
[Pipeline] [slavelnx3] }
[Pipeline] // parallel
[Pipeline] End of Pipeline

```

<p>This is Jenkins 2.32.1 and workflow-cps 2.23</p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> can you confirm this is the issue, or is it just my missing groovy foo?</p>

<p>Would also be happy to hear if someone can come up with a workaround for this</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=289637&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-289637">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Most likely a bug in the depths of <tt>groovy-cps</tt>. Probably easy enough to reproduce in a unit test there. After that, I have no idea offhand how difficult the fix would be.</p>
</pre>
</details>

</ul>
</details>

<p>Most likely a bug in the depths of <tt>groovy-cps</tt>. Probably easy enough to reproduce in a unit test there. After that, I have no idea offhand how difficult the fix would be.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>jelion:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=291522&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-291522">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've been hit by the same issue.  I've noticed that certain kinds of scoping passes through ok, but some kinds do not.</p>

<p>    def say1(s) { println('say1: ' + s) }
     def say2 = { s -&gt; println('say2: ' + s) }</p>

<p>    def setup = {
       def map = <span class="error">&#91;:&#93;</span>
       def say3 = { s -&gt; println('say3: ' + s) }
       for (def i = 0; i &lt; 3; i++) {
         def x = I
         map<span class="error">&#91;x&#93;</span> = { println x; say1( x ); say2( x ); say3( x ) }
       }
       return map
     }</p>

<p>    def jobs = setup()
     parallel(jobs)</p>

<p>The "println" and "say1" work but "say2" and "say3" are wrong, and in a different ways.  Stripping all the "noise", I get the following:</p>

<p>    <span class="error">&#91;0&#93;</span> 0
     <span class="error">&#91;1&#93;</span> 1
     <span class="error">&#91;2&#93;</span> 2
     <span class="error">&#91;0&#93;</span> say1: 0
     <span class="error">&#91;1&#93;</span> say1: 1
     <span class="error">&#91;2&#93;</span> say1: 2
     <span class="error">&#91;0&#93;</span> say2: 2
     <span class="error">&#91;1&#93;</span> say2: 0
     <span class="error">&#91;2&#93;</span> say2: 1
     <span class="error">&#91;0&#93;</span> say3: 2
     <span class="error">&#91;1&#93;</span> say3: 2
     <span class="error">&#91;2&#93;</span> say3: 2</p>

<p>say2 is not printing the x value associated with the parallel branch.  say3 is always printing the x value from branch "2".  (I had originally commented that say2 was working, but it is not.  The branch name in brackets should always match the number at the end of each line.)</p>

<p>Spotting these differences helped me work around the issue and also understanding what the scope resolution issue might be...</p>
</pre>
</details>

</ul>
</details>

<p>I've been hit by the same issue.  I've noticed that certain kinds of scoping passes through ok, but some kinds do not.</p>

<p>    def say1(s) { println('say1: ' + s) }<br/>
     def say2 = { s -> println('say2: ' + s) }</p>

<p>    def setup = {<br/>
       def map = <span class="error">&#91;:&#93;</span><br/>
       def say3 = { s -> println('say3: ' + s) }<br/>
       for (def i = 0; i < 3; i++) {<br/>
         def x = I<br/>
         map<span class="error">&#91;x&#93;</span> = { println x; say1( x ); say2( x ); say3( x ) }<br/>
       }<br/>
       return map<br/>
     }</p>

<p>    def jobs = setup()<br/>
     parallel(jobs)</p>

<p>The "println" and "say1" work but "say2" and "say3" are wrong, and in a different ways.  Stripping all the "noise", I get the following:</p>

<p>    <span class="error">&#91;0&#93;</span> 0<br/>
     <span class="error">&#91;1&#93;</span> 1<br/>
     <span class="error">&#91;2&#93;</span> 2<br/>
     <span class="error">&#91;0&#93;</span> say1: 0<br/>
     <span class="error">&#91;1&#93;</span> say1: 1<br/>
     <span class="error">&#91;2&#93;</span> say1: 2<br/>
     <span class="error">&#91;0&#93;</span> say2: 2<br/>
     <span class="error">&#91;1&#93;</span> say2: 0<br/>
     <span class="error">&#91;2&#93;</span> say2: 1<br/>
     <span class="error">&#91;0&#93;</span> say3: 2<br/>
     <span class="error">&#91;1&#93;</span> say3: 2<br/>
     <span class="error">&#91;2&#93;</span> say3: 2</p>

<p>say2 is not printing the x value associated with the parallel branch.  say3 is always printing the x value from branch "2".  (I had originally commented that say2 was working, but it is not.  The branch name in brackets should always match the number at the end of each line.)</p>

<p>Spotting these differences helped me work around the issue and also understanding what the scope resolution issue might be...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=302400&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-302400">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Makes fix of <a href="https://issue-redirect.jenkins.io/issue/26481" title="Mishandling of binary methods accepting Closure" class="issue-link" data-issue-key="JENKINS-26481"><del>JENKINS-26481</del></a> less usable in practice.</p>
</pre>
</details>

</ul>
</details>

<p>Makes fix of <a href="https://issue-redirect.jenkins.io/issue/26481" title="Mishandling of binary methods accepting Closure" class="issue-link" data-issue-key="JENKINS-26481"><del>JENKINS-26481</del></a> less usable in practice.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/abayer.png" title="abayer's avatar" /> abayer:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=302569&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-302569">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've got an initial PR up at <a href="https://github.com/cloudbees/groovy-cps/pull/61" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/cloudbees/groovy-cps/pull/61</a> that seems to do the trick (i.e., I added a test based on the example given below and over in <a href="https://issue-redirect.jenkins.io/issue/44746" title="Groovy bug when creating Closure in loop" class="issue-link" data-issue-key="JENKINS-44746"><del>JENKINS-44746</del></a>, which showed that <tt>parallel</tt> isn't needed to hit this bug, it failed, I made changes, it passed), but it needs <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke" class="user-hover" rel="kohsuke">kohsuke</a> to weigh in.</p>
</pre>
</details>

</ul>
</details>

<p>I've got an initial PR up at <a href="https://github.com/cloudbees/groovy-cps/pull/61" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/cloudbees/groovy-cps/pull/61</a> that seems to do the trick (i.e., I added a test based on the example given below and over in <a href="https://issue-redirect.jenkins.io/issue/44746" title="Groovy bug when creating Closure in loop" class="issue-link" data-issue-key="JENKINS-44746"><del>JENKINS-44746</del></a>, which showed that <tt>parallel</tt> isn't needed to hit this bug, it failed, I made changes, it passed), but it needs <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jglick" class="user-hover" rel="jglick">jglick</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kohsuke" class="user-hover" rel="kohsuke">kohsuke</a> to weigh in.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/externl.jpg" title="externl's avatar" /> externl:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303002&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303002">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yea, thanks! I'm pretty excited about this fix. I can finally remove all the workarounds! </p>
</pre>
</details>

</ul>
</details>

<p>Yea, thanks! I'm pretty excited about this fix. I can finally remove all the workarounds! </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>reinholdfuereder:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303016&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303016">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks for this fix!</p>

<p>This also allowed to get rid of a workaround in non-parallel usage of closure, where I needed to store closure args in a closure-local variable...</p>
</pre>
</details>

</ul>
</details>

<p>Thanks for this fix!</p>

<p>This also allowed to get rid of a workaround in non-parallel usage of closure, where I needed to store closure args in a closure-local variable...</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>markl_lagendijk:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303018&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303018">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Awesome work!
Now that both this bug and the binary Groovy methods bug are fixed, the experience for pipeline developers will be much better! 
It will spare everyone a lot of, 'Argh! Why isn't this working? What am I doing wrong?'-moments/hours <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>.</p>
</pre>
</details>

</ul>
</details>

<p>Awesome work!<br/>
Now that both this bug and the binary Groovy methods bug are fixed, the experience for pipeline developers will be much better! <br/>
It will spare everyone a lot of, 'Argh! Why isn't this working? What am I doing wrong?'-moments/hours <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/>.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/philmcardlecg.png" title="philmcardlecg's avatar" /> philmcardlecg:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303111&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303111">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Apologies if there's a separate bug for this, but I thought this one would address it too. If I adapt the sample pipeline in the description like so:</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def fn = { val -&gt; println val }

stages = [<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>]

def builders = [:]
<span class="code-keyword">for</span> (stage in stages) {
    builders[stage] = { fn(stage) }
}

parallel builders
</pre>
</div></div>
<p>Based on <a href="https://jenkins.io/doc/pipeline/examples/#parallel-multiple-nodes" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/pipeline/examples/#parallel-multiple-nodes</a></p>

<p>I will see the following output:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>[Pipeline] [a] { (Branch: a)
[Pipeline] [b] { (Branch: b)
[Pipeline] [a] echo
[a] b
[Pipeline] [a] }
[Pipeline] [b] echo
[b] b
[Pipeline] [b] }
</pre>
</div></div>
<p>Is there a separate bug for this then?</p>

<p>At present, I use the workaround of re-assigning anything in the stage object to a local variable before using them, which looked a lot like what was happening here.</p>

<p>If I reintroduce the workaround from the documentation linked, I see the correct output of course.</p>
</pre>
</details>

</ul>
</details>

<p>Apologies if there's a separate bug for this, but I thought this one would address it too. If I adapt the sample pipeline in the description like so:</p>

<pre>
def fn = { val -> println val }

stages = [<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>]

def builders = [:]
<span class="code-keyword">for</span> (stage in stages) {
    builders[stage] = { fn(stage) }
}

parallel builders
</pre>
<p>Based on <a href="https://jenkins.io/doc/pipeline/examples/#parallel-multiple-nodes" class="external-link" target="_blank" rel="nofollow noopener">https://jenkins.io/doc/pipeline/examples/#parallel-multiple-nodes</a></p>

<p>I will see the following output:</p>


```
[Pipeline] [a] { (Branch: a)
[Pipeline] [b] { (Branch: b)
[Pipeline] [a] echo
[a] b
[Pipeline] [a] }
[Pipeline] [b] echo
[b] b
[Pipeline] [b] }

```
<p>Is there a separate bug for this then?</p>

<p>At present, I use the workaround of re-assigning anything in the stage object to a local variable before using them, which looked a lot like what was happening here.</p>

<p>If I reintroduce the workaround from the documentation linked, I see the correct output of course.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>philgrayson:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303112&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303112">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=philmcardlecg" class="user-hover" rel="philmcardlecg">philmcardlecg</a></p>

<p>What you described is not related to this bug. See Jesses' comment from 2017-02-22 22:07.</p>

<p>See also <a href="http://www.teknically-speaking.com/2013/01/closures-in-loops-javascript-gotchas.html" class="external-link" target="_blank" rel="nofollow noopener">http://www.teknically-speaking.com/2013/01/closures-in-loops-javascript-gotchas.html</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=philmcardlecg" class="user-hover" rel="philmcardlecg">philmcardlecg</a></p>

<p>What you described is not related to this bug. See Jesses' comment from 2017-02-22 22:07.</p>

<p>See also <a href="http://www.teknically-speaking.com/2013/01/closures-in-loops-javascript-gotchas.html" class="external-link" target="_blank" rel="nofollow noopener">http://www.teknically-speaking.com/2013/01/closures-in-loops-javascript-gotchas.html</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/philmcardlecg.png" title="philmcardlecg's avatar" /> philmcardlecg:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303115&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303115">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Ah, I hadn't seen that comment. Didn't understand this was a general Groovy issue. Thanks muchly <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>
</pre>
</details>

</ul>
</details>

<p>Ah, I hadn't seen that comment. Didn't understand this was a general Groovy issue. Thanks muchly <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/philmcardlecg.png" title="philmcardlecg's avatar" /> philmcardlecg:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303125&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303125">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>On the off-chance anyone else has my use case, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=philgrayson" class="user-hover" rel="philgrayson">philgrayson</a>'s comment above helped me remember why I wasn't using an iterator, and than I realised that <a href="https://issue-redirect.jenkins.io/issue/26481" title="Mishandling of binary methods accepting Closure" class="issue-link" data-issue-key="JENKINS-26481"><del>JENKINS-26481</del></a> had been fixed in the preceding version of the same plugin, so I'm able to use .each in my example code again and get the desired results (and also in my actual pipeline code).</p>
</pre>
</details>

</ul>
</details>

<p>On the off-chance anyone else has my use case, <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=philgrayson" class="user-hover" rel="philgrayson">philgrayson</a>'s comment above helped me remember why I wasn't using an iterator, and than I realised that <a href="https://issue-redirect.jenkins.io/issue/26481" title="Mishandling of binary methods accepting Closure" class="issue-link" data-issue-key="JENKINS-26481"><del>JENKINS-26481</del></a> had been fixed in the preceding version of the same plugin, so I'm able to use .each in my example code again and get the desired results (and also in my actual pipeline code).</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jglick.png" title="jglick's avatar" /> jglick:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=303143&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-303143">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Right, whereas previously you needed to do</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def builders = [:]
<span class="code-keyword">for</span> (stage in [<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>]) {
    def _stage = stage
    builders[stage] = {echo _stage}
}
parallel builders
</pre>
</div></div>

<p>now you can do</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def builders = [:]
[<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>].each {stage -&gt; builders[stage] = {echo stage}}
parallel builders
</pre>
</div></div>

<p>Actually you can simplify a bit more</p>

<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">parallel([<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>].collectEntries {stage -&gt; [stage, {echo stage}]})
</pre>
</div></div>

<p>though this currently throws up a wall of signature approval requests, some of which are actually internal calls you should not be dealing with; <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> is working on a fix for that.</p>
</pre>
</details>

</ul>
</details>

<p>Right, whereas previously you needed to do</p>


<pre>
def builders = [:]
<span class="code-keyword">for</span> (stage in [<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>]) {
    def _stage = stage
    builders[stage] = {echo _stage}
}
parallel builders
</pre>

<p>now you can do</p>


<pre>
def builders = [:]
[<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>].each {stage -> builders[stage] = {echo stage}}
parallel builders
</pre>

<p>Actually you can simplify a bit more</p>


<pre>
parallel([<span class="code-quote">'a'</span>,<span class="code-quote">'b'</span>].collectEntries {stage -> [stage, {echo stage}]})
</pre>

<p>though this currently throws up a wall of signature approval requests, some of which are actually internal calls you should not be dealing with; <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> is working on a fix for that.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>leedega:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=354376&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354376">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hmmm - I'm not sure if my case is exactly the same as this one, but it appears to me that this bug - or one similar to it - still exists. I'm running Jenkins v2.148 and Pipeline plugin v2.6 and can still break this parallel variable expansion problem, without using closures. Here's an example:</p>

<p> </p>

<p> </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">parallel (
    <span class="code-quote">"thread1"</span>: {
        node() {
            def a = git(branch: <span class="code-quote">'BranchA'</span>, credentialsId: <span class="code-quote">'MyCreds'</span>, url: <span class="code-quote">'url/to/git/repo.git'</span>)
            def b = <span class="code-quote">"1"</span>
            def c = sh(returnStdout: <span class="code-keyword">true</span>, script: <span class="code-quote">'echo 1'</span>).trim() 
            
            echo <span class="code-quote">"git commit "</span> + a.GIT_COMMIT
            echo <span class="code-quote">"b is "</span> + b
            echo <span class="code-quote">"c is "</span> + c
        }
    },    
    <span class="code-quote">"thread2"</span>: {
        node() {
            def a = git(branch: <span class="code-quote">'BranchB'</span>, credentialsId: <span class="code-quote">'MyCreds'</span>, url: <span class="code-quote">'url/to/git/repo.git'</span>)
            def b = <span class="code-quote">"2"</span>
            def c = sh(returnStdout: <span class="code-keyword">true</span>, script: <span class="code-quote">'echo 2'</span>).trim()
            
            echo <span class="code-quote">"git commit "</span> + a.GIT_COMMIT
            echo <span class="code-quote">"b is "</span> + b
            echo <span class="code-quote">"c is "</span> + c
        }
    }
)
</pre>
</div></div>
<p>In this example, I would expect the output to be something like</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>[thread1] git commit &lt;BranchA_Hash&gt;
[thread1] b is 1
[thread1] c is 1
[thread2] git commit &lt;BranchB_Hash&gt;
[thread2] b is 2
[thread2] c is 2</pre>
</div></div>
<p>What I actually get is this:</p>
<div class="preformatted panel" style="border-width: 1px;"><div class="preformattedContent panelContent">
<pre>[thread1] git commit &lt;BranchA_Hash&gt;
[thread1] b is 1
[thread1] c is 1
[thread2] git commit &lt;BranchA_Hash&gt;
[thread2] b is 2
[thread2] c is 2</pre>
</div></div>
<p>While the 'b' and 'c' variables seem to get unique values for each of the parallel threads, the 'a' value becomes ambiguous, always returning a single Git commit hash.</p>

<p>Further, I've confirmed that which Git commit hash that is shown represents the one that is executed "first" in the parallel stages. So, for example, if you put a small "sleep 5" at the top of thread1, then you get the Git commit has from BranchB in both echo statements, and if you move the sleep statement to the top of thread2 you get the has from BranchA.</p>

<p>Also interesting is that although the 'a' variable seems to get confused based on the thread execution, the other variables do not. I thought at first that maybe assigning a static value to the variable may have explained the discrepancy like in the case of the 'b' variable so I added a dynamically executed shell step to the mix for variable 'c' and I get the same results as b... so I'm not sure how / why / what is going on here.</p>
</pre>
</details>

</ul>
</details>

<p>Hmmm - I'm not sure if my case is exactly the same as this one, but it appears to me that this bug - or one similar to it - still exists. I'm running Jenkins v2.148 and Pipeline plugin v2.6 and can still break this parallel variable expansion problem, without using closures. Here's an example:</p>

<p> </p>

<p> </p>

<pre>
parallel (
    <span class="code-quote">"thread1"</span>: {
node() {
    def a = git(branch: <span class="code-quote">'BranchA'</span>, credentialsId: <span class="code-quote">'MyCreds'</span>, url: <span class="code-quote">'url/to/git/repo.git'</span>)
    def b = <span class="code-quote">"1"</span>
    def c = sh(returnStdout: <span class="code-keyword">true</span>, script: <span class="code-quote">'echo 1'</span>).trim() 
    
    echo <span class="code-quote">"git commit "</span> + a.GIT_COMMIT
    echo <span class="code-quote">"b is "</span> + b
    echo <span class="code-quote">"c is "</span> + c
}
    },    
    <span class="code-quote">"thread2"</span>: {
node() {
    def a = git(branch: <span class="code-quote">'BranchB'</span>, credentialsId: <span class="code-quote">'MyCreds'</span>, url: <span class="code-quote">'url/to/git/repo.git'</span>)
    def b = <span class="code-quote">"2"</span>
    def c = sh(returnStdout: <span class="code-keyword">true</span>, script: <span class="code-quote">'echo 2'</span>).trim()
    
    echo <span class="code-quote">"git commit "</span> + a.GIT_COMMIT
    echo <span class="code-quote">"b is "</span> + b
    echo <span class="code-quote">"c is "</span> + c
}
    }
)
</pre>
<p>In this example, I would expect the output to be something like</p>


```
[thread1] git commit <BranchA_Hash>
[thread1] b is 1
[thread1] c is 1
[thread2] git commit <BranchB_Hash>
[thread2] b is 2
[thread2] c is 2
```
<p>What I actually get is this:</p>


```
[thread1] git commit <BranchA_Hash>
[thread1] b is 1
[thread1] c is 1
[thread2] git commit <BranchA_Hash>
[thread2] b is 2
[thread2] c is 2
```
<p>While the 'b' and 'c' variables seem to get unique values for each of the parallel threads, the 'a' value becomes ambiguous, always returning a single Git commit hash.</p>

<p>Further, I've confirmed that which Git commit hash that is shown represents the one that is executed "first" in the parallel stages. So, for example, if you put a small "sleep 5" at the top of thread1, then you get the Git commit has from BranchB in both echo statements, and if you move the sleep statement to the top of thread2 you get the has from BranchA.</p>

<p>Also interesting is that although the 'a' variable seems to get confused based on the thread execution, the other variables do not. I thought at first that maybe assigning a static value to the variable may have explained the discrepancy like in the case of the 'b' variable so I added a dynamically executed shell step to the mix for variable 'c' and I get the same results as b... so I'm not sure how / why / what is going on here.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jekeller.png" title="jekeller's avatar" /> jekeller:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=354392&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354392">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>so you're using the git step, and expect its output to return something unique? Is it possible that after that runs once it always returns the first branch? I suspect maybe that step isn't actually returning the value you think it does?</p>

<p>Hmm, but you said if you rename the variable it works fine?? that's odd.</p>
</pre>
</details>

</ul>
</details>

<p>so you're using the git step, and expect its output to return something unique? Is it possible that after that runs once it always returns the first branch? I suspect maybe that step isn't actually returning the value you think it does?</p>

<p>Hmm, but you said if you rename the variable it works fine?? that's odd.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>leedega:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-38268?focusedId=354396&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-354396">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>So I guess the problem as described in my example above seems to be particular to the `git` and `checkout` build steps, however I have managed to reproduce the problem using the `sh` step as well - just in a more elaborate way. I haven't reduced that one to a minimal reproducible example yet so I left that part out of my reply.</p>

<p>For now the only easy-to-reproduce example I have is with those 2 specific build steps. Also, the "trick" of renaming the variable names doesn't seem to work with either of those two build steps as I had originally thought. It does, however, seem to work around the problem wrt the `sh` build step that I have also encountered in some of my production builds. It's not a "fix" of course but it has temporarily worked around the problem for me in at least one use case. I have others I need to investigate further to see if / how they relate. Will post back as soon as I know more.</p>

<p>Regardless, the example script I provided above should reproduce the problem in that one context at least.</p>
</pre>
</details>

</ul>
</details>

<p>So I guess the problem as described in my example above seems to be particular to the `git` and `checkout` build steps, however I have managed to reproduce the problem using the `sh` step as well - just in a more elaborate way. I haven't reduced that one to a minimal reproducible example yet so I left that part out of my reply.</p>

<p>For now the only easy-to-reproduce example I have is with those 2 specific build steps. Also, the "trick" of renaming the variable names doesn't seem to work with either of those two build steps as I had originally thought. It does, however, seem to work around the problem wrt the `sh` build step that I have also encountered in some of my production builds. It's not a "fix" of course but it has temporarily worked around the problem for me in at least one use case. I have others I need to investigate further to see if / how they relate. Will post back as soon as I know more.</p>

<p>Regardless, the example script I provided above should reproduce the problem in that one context at least.</p>
